### PR TITLE
Basic routing & Authentication

### PR DESCRIPTION
**Pull request by viethung7899**
- Use `app` router
- Authentication & Protected routes
- Multi-step forms
- Styling with TailwindCSS and DaisyUI
- Custom 404 page

### CODE REVIEW

File: app/login/layout.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+export default function RootLayout({
+  children,
+}: {
+  children: React.ReactNode
+}) {
+  const cookieStore = cookies();
+  const token = cookieStore.get("token")?.value;
+
+  if (token) {
```
Comment (chrishappy) [inline app/login/layout.tsx]: Do we want to convert this to a `isLoggedIn()` helper function?

File: app/page.tsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { decodeToken } from "@/lib/user";
+import {cookies} from "next/headers"
+import { redirect } from "next/navigation";
+
+export default function Index() {
+  const token = cookies().get("token")?.value;
+  if (!token) redirect("/login");
```
Comment (chrishappy) [inline app/page.tsx]: `isLoggedIn()` helper function?

File: components/forms/farmer/ChickenFeeding.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    </div>
+
+    <hr className="border-accent my-2 col-span-2" />
+
+    <Label>Closing stock</Label>
+    <PreviewField className="input justify-end" value={closingStock} />
+
+    <Label>Feed per bird</Label>
+    <div className="join">
+      <PreviewField className="w-full join-item input justify-end" value={feedPerBirdInGrams?.toFixed(2)} />
```
Comment (chrishappy) [inline components/forms/farmer/ChickenFeeding.tsx:24]: Changed `.toFixed(2)` to another helper function `formatDrupalFloat()` or something like that?

File: components/forms/farmer/Confirmation.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { useFormContext } from "react-hook-form";
+import Label from "./Label";
+import PreviewField from "./PreviewField";
+import { FarmerJournal } from "@/types/content";
+
+export default function Confirmation() {
+  const {watch} = useFormContext<FarmerJournal>();
+  const data = watch();
```
Comment (chrishappy) [inline components/forms/farmer/Confirmation.tsx:8]: Not a blocker, but just curious:

Is this function working as expected? E.g. allowing us to display derived state?

File: components/forms/farmer/NumberInput.tsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+type Props = {
+  name: FieldPath<FarmerJournal>,
+} & Pick<ComponentProps<'input'>, "className">
+
+export default function NumberInput({ name, className }: Props) {
+  const { register } = useFormContext<FarmerJournal>();
+
+  return <>
+    <input
+      {...register(name, { valueAsNumber: true })}
```
Comment (chrishappy) [inline components/forms/farmer/NumberInput.tsx:14]: Did not know you could do this. Love it!

File: lib/user.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  }
+}
+
+export const decodeToken = (token: string) => {
+  const { iat, drupal } = jwt_decode<JWTPayLoad>(token)
+  return {
+    iat,
+    user: {
+      uid: drupal.uid,
+      role: drupal.role.includes("ovo_technician") ? "TECHNICIAN" : "FARMER",
```
Comment (chrishappy) [inline lib/user.ts:28]: Can we only set users to farmers if they have the role farmers?

Comment (chrishappy) [reply lib/user.ts:28]: Also, I wonder how much authentication DrupalNext is able to handle.

**Review by chrishappy — APPROVED** [review]
Great work, Viet! Everything is look a lot more clean and slick.

I feel like I can breather better without all the `ion` prefixes 😂

Just a few minor suggestions. Other than that, everything looks good to me!

**Comment by viethung7899** [comment]
Form enhancement will be addressed in later PRs

File: app/login/layout.tsx (hunk 1 above)
Comment (viethung7899) [reply app/login/layout.tsx]: Implemented function `getAuthRole()` to get the authorized role from the token

File: components/forms/farmer/Confirmation.tsx (hunk 4 above)
Comment (viethung7899) [reply components/forms/farmer/Confirmation.tsx:8]: It works as expected, but has nothing to do with the computed states.

**Comment by chrishappy** [comment]
Great job, Viet! Exciting!!!

**Pull request merged**